### ISSUE TITLE
Gradient called twice per iterations.

### ISSUE THREAD

**Issue by KristofferC** [EconForge/NLsolve.jl#32]
As it is right now the gradient function is called twice per iteration in the newton scheme.

The first one here: https://github.com/EconForge/NLsolve.jl/blob/ae1e89a5795324a7a35f82a22499932cc1f569f7/src/newton.jl#L96

and the second one in the linesearch i.e: https://github.com/JuliaOpt/Optim.jl/blob/master/src/linesearch/backtracking_linesearch.jl#L26

This is a bit unfortunate since the gradient is sometimes quite expensive to compute.

We could fix this by "cheating" by knowing that the linesearches only computes the gradients at the current point where we have already computed the derivative.

As an example, consider this test case with an artificial sleep introduced in the gradient:

```jl
function f!(x, fvec)
    fvec[1] = (x[1]+3)*(x[2]^3-7)+18
    fvec[2] = sin(x[2]*exp(x[1])-1)
end

function g!(x, fjac)
    fjac[1, 1] = x[2]^3-7
    fjac[1, 2] = 3*x[2]^2*(x[1]+3)
    u = exp(x[1])*cos(x[2]*exp(x[1])-1)
    fjac[2, 1] = x[2]*u
    fjac[2, 2] = u
    sleep(1)
end

df = DifferentiableMultivariateFunction(f!, g!)
```

I get this on master (7 seconds):
```jl
julia> @time r = nlsolve(df, [ -0.5; 1.4], method = :newton, linesearch! = Optim.backtracking_linesearch!, ftol = 1e-6)
  7.020818 seconds (237 allocations: 10.500 KB)
Results of Nonlinear Solver Algorithm
 * Algorithm: Newton with line-search
 * Starting Point: [-0.5,1.4]
 * Zero: [-1.779077697780431e-8,1.0000000055513854]
 * Inf-norm of residuals: 0.000000
 * Iterations: 3
 * Convergence: true
   * |x - x'| < 0.0e+00: false
   * |f(x)| < 1.0e-06: true
 * Function Calls (f): 10
 * Jacobian Calls (df/dx): 6
end
```

Using #33 I get (3 seconds):

```jl
julia> @time r = nlsolve(df, [ -0.5; 1.4], method = :newton, linesearch! = Optim.backtracking_linesearch!, ftol = 1e-6)
  3.009181 seconds (189 allocations: 8.359 KB)
Results of Nonlinear Solver Algorithm
 * Algorithm: Newton with line-search
 * Starting Point: [-0.5,1.4]
 * Zero: [-1.779077697780431e-8,1.0000000055513854]
 * Inf-norm of residuals: 0.000000
 * Iterations: 3
 * Convergence: true
   * |x - x'| < 0.0e+00: false
   * |f(x)| < 1.0e-06: true
 * Function Calls (f): 10
 * Jacobian Calls (df/dx): 3
end
```

It is a bit ugly and we have to double check that it works with the other line searches but it is my opinion that we need to do something to fix this because double calling the graident is unacceptable.

**Comment by KristofferC** [comment]
Thoughts @timholy @spencerlyon2 ?

**Comment by KristofferC** [comment]
This works well for the backtracking linesearch but the other searches seem to recompute the gradient at different points.

I thought a line search only used the gradient information at the trial point for the current iteration...?

**Comment by KristofferC** [comment]
Another way could be to check if the point `x` is equal to the iteration point and only in that case give the already computed gradient. This should fall back nicely to the cases where the gradient need to be computed on a new point in the linesearch. The cost for the check for x being equal should be insginificant.

**Comment by KristofferC** [comment]
Implemented my last comment in #33
